k8s/cmd/coredns/plugin/fanout/fanout.go

```diff
@@ -0,0 +1,108 @@
+package fanout
+
+import (
+	"context"
+	"crypto/tls"
+	"time"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/debug"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+var log = clog.NewWithPlugin("fanout")
+
+// Fanout represents makeRecordA plugin instance that can do async proxy requests to another (DNS) servers.
+type Fanout struct {
+	agents []*dnsAgent
+
+	tlsConfig     *tls.Config
+	tlsServerName string
+	failLimit     int
+	expire        time.Duration
+
+	Next plugin.Handler
+}
+
+type connectResult struct {
+	server   *dnsAgent
+	response *dns.Msg
+	start    time.Time
+}
+
+// NewFanout returns reference to new Fanout plugin instance with default configs.
+func NewFanout() *Fanout {
+	f := &Fanout{
+		failLimit: 2,
+		tlsConfig: new(tls.Config),
+		expire:    defaultExpire,
+	}
+	return f
+}
+
+// setProxy appends p to the proxy list and starts healthchecking.
+func (f *Fanout) setProxy(p *dnsAgent) {
+	f.agents = append(f.agents, p)
+	p.start(healthClientInterval)
+}
+
+// Len returns the number of configured agents.
+func (f *Fanout) Len() int {
+	return len(f.agents)
+}
+
+// Name implements plugin.Handler.
+func (f *Fanout) Name() string {
+	return "fanout"
+}
+
+// ServeDNS implements plugin.Handler.
+func (f *Fanout) ServeDNS(ctx context.Context, w dns.ResponseWriter, m *dns.Msg) (int, error) {
+	req := request.Request{W: w, Req: m}
+	result := make(chan connectResult, len(f.agents))
+	for i, _ := range f.agents {
+		agent := f.agents[i]
+		go func() {
+			start := time.Now()
+			for attempt := 0; attempt < f.failLimit; attempt++ {
+				resp, err := agent.Connect(ctx, req)
+				if err == errCachedClosed {
+					continue
+				}
+				if err == nil && len(resp.Answer) != 0 {
+					result <- connectResult{
+						server:   agent,
+						response: resp,
+						start:    start,
+					}
+					return
+				}
+				if err != nil {
+					agent.healthcheck()
+				}
+			}
+		}()
+	}
+	var first connectResult
```

Comment:
This is less a comment on the code and more a question on the spec.   If I am reading correctly the local coredns server in the sidecar always has strict priority over requests since the resolv.conf always points first to it then to the kube-dns server.   So if there was ever a case where a service with the same name was registered in both k8s and in the DNS server the connection points to then the server from the NSM connection would provide the response.   Is this the correct understanding?

Reply:
>If I am reading correctly the local coredns server in the sidecar always has strict priority over requests since the resolv.conf always points first to it then to the kube-dns server. 
Actually, we doing request only to local coredns. It fanout request to all known nameservers inclues default k8s name server. The first positive response is answer to request.

Reply:
Ok looking closer I see that the implemetnation of replaceDNSconfig doesn't match what was described in the doc.   There probably needs to be some deterministic way to decide if the DNS form the connections or DNS from K8s should take priority if both return matches.   The K8s DNS is automatic and not easily switched off IMO it is problematic if there isn't some way to specify priority vs. the first server to respond.   If I am missing some aspect let me know. 

Sine this is closed I guess I need to repeat in another PR when I see it

Reply:
So, as I understand the problem is next scenario:

1) local coredns know about two DNS Servers
1.1) DNS Server1: default k8s DNS server
1.2) DNS Server2: user's DNS server
2) DNS Server1 and DNS Server2 know about the host with name `www.google.com`
3) The user wants that his DNS Server always answers the question related to `www.google.com`
4) In this case, the user just simple add search domain to your `DnsConfig` (look at the proto message into Spec)
It will generate next corefile
```
. {
	fanout {Default k8s DNS Server IP} {User's DNS Server IP}
}
www.google.com {
	hosts {
		 {User's DNS Server IP} www.google.com
	}
}
```
This config will by priority sents all queries about `www.google.com` to {User's DNS Server IP}

Reply:
Let me know if I missed something